nixos/modules/services/x11/dunst.nix

```diff
@@ -0,0 +1,289 @@
+    default = defaultFrameConfig;
+    description = ''
+      The frame configuration for dunst. Values will be merged into the defaults.
+    '';
+    example = ''
+      {
+        width = 0;
+        color = "#000000";
+      }
+    '';
+  };
+
+  urgencyConfig = {
+    low = mkOption {
+      type = with types; attrsOf (oneOf [bool str int]);
+      default = defaultUrgencyConfigLow;
+      description = ''
+        The low urgency section of the dunst configuration. Values will be merged into the defaults.
+      '';
+      example = ''
+        {
+          background = "#ffffff";
+          foreground = "#888888";
+          timeout = 60;
+        }
+      '';
+    };
+    normal = mkOption {
+      type = with types; attrsOf (oneOf [bool str int]);
+      default = defaultUrgencyConfigNormal;
+      description = ''
+        The normal urgency section of the dunst configuration. Values will be merged into the defaults.
+      '';
+      example = ''
+        {
+          background = "#ffffff";
+          foreground = "#111111";
+          timeout = 60;
+        }
+      '';
+    };
+    critical = mkOption {
+      type = with types; attrsOf (oneOf [bool str int]);
+      default = defaultUrgencyConfigCritical;
+      description = ''
+        The critical urgency section of the dunst configuration. Values will be merged into the defaults.
+      '';
+      example = ''
+        {
+          background = "#ff0000";
+          foreground = "#111111";
+          timeout = 0;
+        }
+      '';
+    };
+  };
+
+  experimentalConfig = mkOption {
+    type = with types; attrsOf (oneOf [bool str int]);
+    default = defaultExperimentalConfig;
+    description = ''
+      The experimental configuration for dunst. Values will be merged into the defaults.
+    '';
+    example = ''
+      {
+        per_monitor_dpi = true;
+      }
+    '';
+  };
+
+  rules = mkOption {
+    type = with types; attrsOf (oneOf [bool str int]);
+    default = {};
+    description = ''
+       These rules allow the conditional modification of notifications.
+
+       Note that rule names may not be one of the following
+       keywords already used internally:
+         ${concatStringsSep ", " reservedSections}
+       There are 2 parts in configuring a rule: Defining when a rule
+       matches and should apply (called filtering in the man page)
+       and then the actions that should be taken when the rule is
+       matched (called modifying in the man page).
+    '';
+    example = literalExample ''
+      signed_off = {
+        appname = "Pidgin";
+        summary = "*signed off*";
+        urgency = "low";
+        script = "pidgin-signed-off.sh";
+      };
+    '';
+  };
+
+};
+
+config =
+  let
+    dunstValueString = v: with builtins;
+      let err = t: v: abort
+        ("generators.mkValueStringDefault: " +
+         "${t} not supported: ${toPretty {} v}");
+      in   if isInt      v then toString v
```

Comment:
Can't you use builtins.toJSON here? It seems to give identical results.

Reply:
Are you asking if 
1. I can replace `dunstValueString` with `builtins.toJSON`, or
2. just `toString v`? 

For 1, I think `builtins.toJSON` is accepting of more values than `dunstValueString`. For example, I could set a config value to a list `[1 2 3]` and it would build to `[1, 2, 3]` which isn't a sensible dunst config value. Being more accepting I think would cause user confusion.

For 2. I could do that and I wouldn't have the accepting issues of 1 because of the `if isInt v`, but what does `builtins.toJSON` add over `builtins.toString`.

If you're asking something different and I've misunderstood, I apologize for the confusion. Though I do thank you for pointing out this section of code because line 243 should read `dunstValueString` not `generators.mkValueStringDefault`. I'll fix that.

Reply:
I meant something like:
```
dunstValueString = v: with builtins;
  let
    err = t: v: abort
      ("generators.mkValueStringDefault: " +
       "${t} not supported: ${toPretty {} v}");

    isValid v = any id [(isInt v) (isString v) v];

  in if isValid v
       then builtins.toJSON v
       else err "this value is" (toString v);
```
You're still doing the same type-checking but reusing the serialisation, which seems to be identical to JSON, in these cases.

Reply:
I see now. Thanks for the example I'll incorporate this.